torch/lib/c10d/TCPStore.cpp

```diff
@@ -591,6 +591,229 @@ void TCPStoreMasterDaemon::run() {
 }
 #endif
 
+namespace detail {
+namespace {
+
+// Offers RAII for TCP sockets.
+class TCPSocket {
+ public:
+  TCPSocket() noexcept = default;
+
+  /* implicit */ TCPSocket(int handle) noexcept : handle_{handle} {}
+
+  TCPSocket(const TCPSocket& other) = delete;
+
+  TCPSocket& operator=(const TCPSocket& other) = delete;
+
+  TCPSocket(TCPSocket&& other) noexcept : handle_{other.handle_} {
+    other.handle_ = c10::nullopt;
+  }
+
+  TCPSocket& operator=(TCPSocket&& other) noexcept {
+    closeSocket();
+
+    handle_ = std::exchange(other.handle_, c10::nullopt);
+
+    return *this;
+  }
+
+  ~TCPSocket() {
+    closeSocket();
+  }
+
+  int handle() const noexcept {
+    return handle_.value_or(-1);
+  }
+
+ private:
+  void closeSocket() noexcept {
+    if (handle_) {
+      tcputil::closeSocket(*handle_);
+    }
+  }
+
+  c10::optional<int> handle_{};
+};
+
+} // namespace
+
+// Manages the lifecycle of a server daemon.
+class TCPServer {
+ public:
+  static std::shared_ptr<TCPServer> start(const TCPStoreOptions& opts);
+
+  PortType port() const noexcept {
+    return port_;
+  }
+
+  explicit TCPServer(
+      TCPSocket&& socket,
+      PortType port,
+      std::unique_ptr<TCPStoreMasterDaemon>&& daemon)
+      : socket_{std::move(socket)}, port_{port}, daemon_{std::move(daemon)} {}
+
+ private:
+  TCPSocket socket_;
+  PortType port_;
```

Comment:
can these be const?

Reply:
Technically yes, however, I usually avoid using `const` qualifier for member variables as they interfere with move semantics (compiler silently fallbacks to copy semantic due to "constness" of a variable). In this particular case it should be safe though as we always heap allocate `TCPServer` and move it around with a `std::unique_ptr<>`. I will make them `const` in the next revision.